server/extension/extension.go

```diff
@@ -0,0 +1,393 @@
+		return nil, err
+	}
+	proxy := httputil.NewSingleHostReverseProxy(url)
+	proxy.Transport = newTransport(config)
+	return proxy, nil
+}
+
+// newTransport will build a new transport to be used in the proxy
+// applying default values if not defined in the given config.
+func newTransport(config ProxyConfig) *http.Transport {
+	applyProxyConfigDefaults(&config)
+	return &http.Transport{
+		DialContext: (&net.Dialer{
+			Timeout:   config.ConnectionTimeout,
+			KeepAlive: config.KeepAlive,
+		}).DialContext,
+		MaxIdleConns:          config.MaxIdleConnections,
+		IdleConnTimeout:       config.IdleConnectionTimeout,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+	}
+}
+
+func applyProxyConfigDefaults(c *ProxyConfig) {
+	if c.ConnectionTimeout == 0 {
+		c.ConnectionTimeout = DefaultConnectionTimeout
+	}
+	if c.KeepAlive == 0 {
+		c.KeepAlive = DefaultKeepAlive
+	}
+	if c.IdleConnectionTimeout == 0 {
+		c.IdleConnectionTimeout = DefaultIdleConnectionTimeout
+	}
+	if c.MaxIdleConnections == 0 {
+		c.MaxIdleConnections = DefaultMaxIdleConnections
+	}
+}
+
+// RegisterHandlers will retrieve all configured extensions
+// and register the respective http handlers in the given
+// router.
+func (m *Manager) RegisterHandlers(r *mux.Router) error {
+	m.log.Info("Registering extension handlers...")
+	config, err := m.settings.Get()
+	if err != nil {
+		return fmt.Errorf("error getting settings: %s", err)
+	}
+
+	if config.ExtensionConfig == "" {
+		return fmt.Errorf("No extensions configurations found")
+	}
+
+	extConfigs, err := parseConfig(config.ExtensionConfig)
+	if err != nil {
+		return fmt.Errorf("error parsing extension config: %s", err)
+	}
+	return m.registerExtensions(r, extConfigs)
+}
+
+// registerExtensions will iterate over the given extConfigs and register
+// http handlers for every extension. It also registers a list extensions
+// handler under the "/extensions/" endpoint.
+func (m *Manager) registerExtensions(r *mux.Router, extConfigs *ExtensionConfigs) error {
+	extRouter := r.PathPrefix(fmt.Sprintf("%s/", URLPrefix)).Subrouter()
+	for _, ext := range extConfigs.Extensions {
+		proxyByCluster := make(map[string]*httputil.ReverseProxy)
+		for _, service := range ext.Backend.Services {
+			proxy, err := NewProxy(service.URL, ext.Backend.ProxyConfig)
+			if err != nil {
+				return fmt.Errorf("error creating proxy: %s", err)
+			}
+			proxyByCluster[service.ClusterName] = proxy
+		}
+		m.log.Infof("Registering handler for %s/%s...", URLPrefix, ext.Name)
+		extRouter.PathPrefix(fmt.Sprintf("/%s/", ext.Name)).
+			HandlerFunc(m.CallExtension(ext.Name, proxyByCluster))
+	}
+	// register /extensions/ endpoint to list all configured extensions
+	extRouter.HandleFunc("/", m.ListExtensions(extConfigs))
+	return nil
+}
+
+// ListExtensions returns a handler func to list all configured extensions
+func (m *Manager) ListExtensions(extConfigs *ExtensionConfigs) func(http.ResponseWriter, *http.Request) {
```

Comment:
My main idea was to provide a tool for the ArgoCD admin to provide info about which extensions are correctly registered in the proxy. This is a configuration that comes from a yaml string in the configmap and can be easily malformated. However, this is not part of the proposal and I added it as it was helping me to test and validate the proxy config and thought that it would be useful for others too. I could enforce this endpoint behind an admin rbac. If you feel strong about this I can remove it from the code.
WDYT?

Reply:
I feel like exposing any endpoint without a really solid use case and first-class features consuming it (UI, CLI), we should just drop it.

If we need a first-class feature for debugging, we should just add that.

Reply:
removed